Update migration guide v3-> v4 (#794)
LiskHQ/lisk-core: docs/antora/modules/ROOT/pages/management/migration.adoc

= Testnet and Mainnet v4 migration guide
:toc: preamble
:experimental:
:idprefix:
:idseparator: -
:snapshotHeight: TBD
:snapshotHeightTestnet: TBD

> :initRounds: TBD

Author: Done: https://github.com/LiskHQ/lisk-core/issues/795

> :lisk-migrator: lisk-migrator-v2.0.0-linux-x64.tar.gz

Reviewer: I think linux-x64, darwin-x64 and darwin-arm64

====

:sectnums:
:sectnumlevels: 5
== Preparation

.System requirements
[NOTE]

====
The following system requirements are recommended for running the migration script:

Memory::
* minimum of 4 GB RAM

OS::
* Ubuntu 20.04 LTS
* Ubuntu 22.04 LTS

Node.js::
* Version 18 (LTS)
====

=== Preparing Lisk Core v3.0.4

==== Ensure you are running the binary distribution of Lisk Core
If you are using a different distribution of Lisk Core v3.0.4, such as Docker, Commander or Source, it will be necessary to install the application/binary distribution.

==== Ensure you are running version 3.0.4 of Lisk Core
Ensure you are running version 3.0.4 of Lisk Core to be able to seamlessly migrate to Lisk Core 4.0.0.

Open `package.json` to check the `version` of your Lisk Core node.
If your version is lower than 3.0.4, ensure to update Lisk Core before proceeding with the migration.

=== Setting up the Lisk migrator

==== Download the migration script
Download the migration script by running the following command in the terminal:

[subs=attributes+]
[source,bash]
----
curl -o {lisk-migrator} {url_lisk_migrator}
----

> ==== Download checksum and verify

Reviewer 2: ```suggestion
==== Download the checksum and verify
```

> Download the checksum and verify the successful download of lisk-migrator.

Reviewer 2: ```suggestion
Download the checksum and verify the successful download of the lisk-migrator.
```

{counter:seq3:A}) Download the checksum.

[subs=attributes+]
[source,bash]
----
curl -o {lisk-migrator}.SHA256 {url_lisk_migrator}.SHA256
----

{counter:seq3}) Run the following command in the terminal and ensure the output is `sha256sum: <file name>: OK`

[subs=attributes+]
[source,bash]
----

> sha256sum -c {lisk-migrator}.SHA256

Reviewer 3: Darwin doesn't offer `sha256sum` like in Linux. Instead, you'll need to use `sha -a 256`.

----

==== Extract and add to PATH

Unpack it, and then add it to the system path, in order to use it in the terminal:

[subs=attributes+]
[source,bash]
----
tar -xf ./{lisk-migrator}
----

Make the `lisk-migrator` command available in the PATH, e.g. by executing the following command:

[source,bash]
----
export PATH="$PATH:$HOME/lisk-migrator/bin"
----

> In case `lisk-migrator` was extracted somewhere else other than in your home directory, replace `$HOME` with the absolute path of where the `lisk-migrator` folder is located.

Reviewer 2: ```suggestion
In case the `lisk-migrator` was extracted somewhere else other than in your home directory, replace `$HOME` with the absolute path of where the `lisk-migrator` folder is located.
```

== Migration steps

=== Check the announced snapshot height

* For Mainnet: {snapshotHeight} +
For more information check the {url_lisk_blog_migration_mainnet}[Mainnet migration announcement^].
* For Testnet: {snapshotHeightTestnet} +

> For more information check the {url_lisk_blog_migration}[Testnet migration announcement^].

Author: yes, this needs to be updated once the articles are out. I can comment it out for now

> The height is needed by `lisk-migrator` in the next step.

Reviewer 2: ```suggestion
The height is needed by the `lisk-migrator` in the next step.
```

A snapshot of the blockchain will be created at this particular height, which will then be used to create the genesis block for the new blockchain.

=== Ensure Lisk Core v3.0.4 is fully synced with the network

> Check the current block height of your node directly in the terminal by running the following command:

Author: This is covered in the preparation steps :)

[source,bash]
----
lisk-core node:info --pretty
----

Compare the current height of your node to the network height in {url_wallet}[Lisk Desktop^], which is shown on the kbd:[Network] or kbd:[Blocks] pages.

If both heights are equal, it is verified that your node is fully synced with the network.

TIP: To view the current height of the *Lisk Testnet*, use the network switcher of Lisk Desktop, which can be enabled in the settings.

=== Run lisk migrator

[IMPORTANT]
====
.When to start the migrator script?
`lisk-migrator` can be started any time before the announced snapshot height.
====

> If you have added `lisk-migrator` to the PATH as described in the section <<setting-up-the-lisk-migrator>>, you can start the migration script by running the following command footnote:snap_footnote[Snap versions of Lisk Core store everything in `~/snap/lisk-core/current/.lisk/lisk-core` instead of `~/.lisk/lisk-core`] in the terminal:

Reviewer 2: ```suggestion
If you have added the `lisk-migrator` to the PATH as described in the section <<setting-up-the-lisk-migrator>>, you can start the migration script by running the following command footnote:snap_footnote[Snap versions of Lisk Core store everything in `~/snap/lisk-core/current/.lisk/lisk-core` instead of `~/.lisk/lisk-core`] in the terminal:
```

----
--
Testnet::
+
--
[source,bash,subs=attributes+]
----
lisk-migrator --snapshot-height {snapshotHeightTestnet} --output ~/.lisk/lisk-core/config/testnet --lisk-core-v3-data-path ~/lisk-test --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
----
--
====

* `--snapshot-height`:
The height on which the blockchain snapshot will be performed.
The snapshot height will be announced separately.
* `--output`:
The absolute path to the directory, where the newly generated genesis block should be saved.
* `--lisk-core-v3-data-path`:
The absolute path to the directory, where the Lisk Core v3.0.4 node is located.

[TIP]

====
It is possible to use tools such as `screen` to run the Lisk migrator in the background.

With `screen` you can detach the current terminal window into the background:

.Example (Mainnet) footnote:snap_footnote[]
[source,bash,subs=attributes+]
----
screen -dmSL migration lisk-migrator --snapshot-height {snapshotHeight} --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/lisk-main --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
----

> Shortly before the migration occurs, it is possible to reattach to it in order to check if everything is working correctly.

Reviewer 2: ```suggestion
Shortly before the migration occurs, it is possible to reattach the screen to it in order to check if everything is working correctly.
```

Author: ```suggestion
Shortly before the migration occurs, it is possible to reattach to the screen, in order to check if everything is working correctly.
```

> First check the name of the detached screen:

Reviewer 2: ```suggestion
First, check the name of the detached screen:
```

[source,bash]
----
screen -ls
----

This returns a list of all detached screens with `screen`:

----
There is a screen on:
	1842.migration	(05/07/2021 12:35:59 PM)	(Detached)
1 Socket in /run/screen/S-lisk.
----

Use `screen -r` and the name of the detached screen you want to connect to

[source,bash]
----
screen -r 1842.migration
----
====

=== Wait until the network reaches the snapshot height

> Observe if `lisk-migrator` finishes successfully, (this can take several minutes).

Reviewer 2: ```suggestion
Observe if the `lisk-migrator` finishes successfully, (this can take several minutes).
```

The flag `--auto-download-lisk-core-v4` will also download and install Lisk Core v4 for you.

> The flag `--auto-migrate-config` will automatically migrate the config from old to new node.

Reviewer 2: ```suggestion
The flag `--auto-migrate-config` will automatically migrate the config from the old to the new node.
```

//TODO: Update timeframe for initRounds

> After the snapshot height is reached, validators have approximately 2 hours time to enable block generation, to ensure they will not miss any blocks after the hardfork.

Reviewer 2: ```suggestion
After the snapshot height is reached, validators have approximately 2 hours time to enable block generation, to ensure they will not miss any blocks after the hard fork.
```

You can verify that the node is running correctly by executing the following command:

[source,bash]
----
lisk-core system:node-info --pretty
----

This will return certain general node information.

=== Stop Lisk Core v3.0.4

After the `lisk-migrator` script has finished and the announced snapshot height has passed, there is no reason to continue running Lisk Core v3.0.4 and therefore it is recommended to stop it.

First ensure, that Lisk Core will not start again by removing the existing cronjob:

[source,bash]
----
crontab -e
----

Choose your favorite editor to open the `crontab` file and remove the following line:

 @reboot /bin/bash /home/lisk/lisk-test/lisk.sh start > /home/lisk/lisk-test/cron.log 2>&1

Save and close the `crontab` file again.

Navigate into the root folder of your Lisk Core v3.0.4 installation and run the following command to stop the old Lisk Core version:

[source,bash]
----
bash lisk.sh stop
----

Last but not least, remove the folder with Lisk Core v3.0.4, e.g. by executing the following:

[tabs]
====
Mainnet::
+
--
[source,bash]
----
rm -r lisk-main
----
--
Testnet::
+
--
[source,bash]
----
rm -r lisk-test
----
--
====

> === Enabling block generation after migration (for validators)

Reviewer 2: ```suggestion
=== Enabling block generation after the migration (for validators)
```

After migration, *101 initial validators* will be active to generate blocks during the `initRounds`.

> `initRounds` is number of rounds for the bootstrap period of the new network.

Reviewer 2: ```suggestion
`initRounds` is the number of rounds for the bootstrap period of the new network.
```

Without the hash onion, a validator won't be able to receive any rewards for generating new blocks, although the blocks would still be valid in that case.
To not miss any rewards, it is of high interest for a validator to set the hash onion, before enabling block generation on the node.

[source,bash]
----
lisk-core endpoint:invoke random_setHashOnion '{"address":"lskqaxxmj78frvgpjgwvf4yqjjkcrr9yhn2sxxwm3"}'
----

[CAUTION]
====
If you see the following error after running the command:

[source,bash]
----
$ lisk-core endpoint:invoke random_setHashOnion '{"address":"lskxsddtvbc7ze98bqbawaq3ydj2f8387429633t9"}'
 ›   Error: Response not received in 3000ms
----

This suggests that the API timed out.
However, the hash onion is still created in most cases.
You can wait for a while and then get the hash onion, to verify that it actually is set.

[source,bash]
----
lisk-core endpoint:invoke random_getHashOnionSeeds --pretty
----
====

==== Enable block generation

Now, it is possible to enable block generation on the new node for your validator by using the Lisk Core CLI.

[source,bash]
----
lisk-core generator:enable lskqaxxmj78frvgpjgwvf4yqjjkcrr9yhn2sxxwm3 --use-status-value
----

> Replace the address with your delegate address in hexadecimal representation.

Author: ```suggestion
Replace the address with your validator address.
```

= Testnet and Mainnet v4 migration guide
:toc: preamble
:experimental:
:idprefix:
:idseparator: -
:snapshotHeight: TBD
:snapshotHeightTestnet: TBD
:initRounds: TBD
:lisk-migrator: lisk-migrator-v2.0.0-linux-x64.tar.gz
:docs_sdk: v6@lisk-sdk::
// External URLs
:url_lisk_chat: https://lisk.chat/
:url_lisk_migrator: https://downloads.lisk.com/lisk-migrator/{lisk-migrator}
:url_wallet: https://lisk.com/wallet
:url_jsonrpc: https://www.jsonrpc.org/specification

> :url_lisk_blog_migration: https://lisk.com/blog/development/announcing-lisk-testnet-v3-migration

Author: It are links to the old migration posts, they are commented out in the content, and will be updated and uncommented, once the new blog posts are published :)

:url_lisk_blog_migration_mainnet: https://lisk.com/blog/development/announcing-lisk-mainnet-v3-migration
// Project URLs
:url_run_validator: ROOT::run-blockchain/become-validator.adoc
:url_sdk_cli_keyscreate: {docs_sdk}client-cli.adoc#keyscreate
:url_setup_binary: swtup/binary.adoc

This guide explains how to migrate a Lisk Core v3.0.4 node to Lisk Core v4.x using the Lisk migrator v2.

The Lisk migrator CLI will generate a new genesis block for Lisk Core v4.
The new genesis block is created based on a snapshot of the old blockchain (running on Lisk Core v3.0.4) at a given height.

[IMPORTANT]

====
All active validators on the Lisk Mainnet and the Lisk Testnet need to follow this guide to correctly migrate their nodes to the new network, in order to not miss any blocks after the network hard fork.

Optionally, anyone running a Lisk Core v3.0.4 who is not an active validator can participate in the migration process.

> //TODO: Update steps

Author: resolved it 👍

[source,bash,subs=attributes+]
----
screen -dmSL migration lisk-migrator --snapshot-height {snapshotHeight} --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/lisk-main --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
----

Shortly before the migration occurs, it is possible to reattach to the screen, in order to check if everything is working correctly.

First, check the name of the detached screen:

[source,bash]
----
screen -ls
----

This returns a list of all detached screens with `screen`:

----
There is a screen on:
	1842.migration	(05/07/2021 12:35:59 PM)	(Detached)
1 Socket in /run/screen/S-lisk.
----

Use `screen -r` and the name of the detached screen you want to connect to

[source,bash]
----
screen -r 1842.migration
----
====

=== Wait until the network reaches the snapshot height

Observe if the `lisk-migrator` finishes successfully, (this can take several minutes).

The flag `--auto-download-lisk-core-v4` will also download and install Lisk Core v4 for you.
The flag `--auto-migrate-config` will automatically migrate the config from the old to the new node.

//TODO: Update timeframe for initRounds

> After the snapshot height is reached, validators have approximately 2 hours time to enable block generation, to ensure they will not miss any blocks after the hard fork.

Reviewer 4: I think this timeline of 2hrs is not measured? maybe we can put this stats based on testnet/mainnet migration testing?

        "mac": "2b3c65d0385a870ab499dfcddf411347506671015f412b35600153b132a455ea",
        "kdf": "argon2id",
        "kdfparams": {
          "parallelism": 4,
          "iterations": 1,
          "memorySize": 2024,
          "salt": "a5598628001346f608b3f57dd38b8611"
        },
        "cipher": "aes-256-gcm",
        "cipherparams": {
          "iv": "683600a199d154e51c0f97e6",
          "tag": "32807058f7f89921b4839fc39256cd24"
        },
        "version": "1"
      }
    }
  ]
}
----

[CAUTION]
====
The generated keys are very sensitive information.

Especially the non encrypted values need to be treated *as equally sensitive as the passphrase for an account.*
So after the keys are imported in step <<import-the-validator-keys>>, make sure to store the file somewhere safe, or delete the file completely.
====

==== Import the validator keys

After creating the validator keys as suggested in the  <<create-the-validator-keys>> section, the next step is to import them into the node.

[source,bash]
----
lisk-core keys:import --file-path config/keys.json
----

> ==== Set the hash onion

Reviewer 4: same here, where possible we can link the existing docs? or it was by choice that we have it separate?

Without the hash onion, a validator won't be able to receive any rewards for generating new blocks, although the blocks would still be valid in that case.
To not miss any rewards, it is of high interest for a validator to set the hash onion, before enabling block generation on the node.

[source,bash]
----
lisk-core endpoint:invoke random_setHashOnion '{"address":"lskqaxxmj78frvgpjgwvf4yqjjkcrr9yhn2sxxwm3"}'
----

[CAUTION]
====
If you see the following error after running the command:

[source,bash]
----
$ lisk-core endpoint:invoke random_setHashOnion '{"address":"lskxsddtvbc7ze98bqbawaq3ydj2f8387429633t9"}'
 ›   Error: Response not received in 3000ms
----

This suggests that the API timed out.
However, the hash onion is still created in most cases.
You can wait for a while and then get the hash onion, to verify that it actually is set.

[source,bash]
----
lisk-core endpoint:invoke random_getHashOnionSeeds --pretty
----
====

> ==== Enable block generation

Reviewer 4: cant we link to the sections in docs https://lisk.com/documentation/lisk-core/v4/core-cli.html#generatorenable?

All other validators will be banned during the bootstrap period.

`initRounds` is the number of rounds for the bootstrap period of the new network.
The bootstrap period after migration to Core v4 is {initRounds} rounds.

The initial validators will be exactly the 101 validators that were in active positions in the Lisk Core v3 network, at the time of the migration.

For the initial validators, it will be important to enable block generation as soon as possible on the new node, to not miss any block rewards.

It is also important that as many as possible validators participate in the network migration.
If only a small number of validators migrate their nodes, this can prolong the bootstrap period for a long time.
Additionally, for blocks to be finalized, enough of the active validators (minimum 68) need to participate in the migration process.

After {initRounds} rounds, the normal Lisk PoS protocol will be followed to generate the list of active validators, but only for the validators who have <<register-legacy-validator-keys,registered their validator keys>>.

WARNING: All validators who *did not register their keys* until the bootstrap period ends will stay *banned*.

> To be un-banned, a validator needs to <<register-legacy-validator-keys>>.

Reviewer 3: The users are unaware so far that the migration process marks all the non-init validators `banned`. So, maybe already give them that context here or in the paragraphs before.

Author: I aded this info in line 392 :)

Now, it is possible to enable block generation on the new node for your validator by using the xref:{url_cli_generatorenable}[generator:enable] command of the Lisk Core CLI.

[source,bash]
----
lisk-core generator:enable lskqaxxmj78frvgpjgwvf4yqjjkcrr9yhn2sxxwm3 --use-status-value
----

Replace the address with your validator address.

.Don't use zeros as validator info data!
[WARNING]
====
If the validator already generated blocks with Lisk Core v3, they need to use their current validator info data.
The validator info data is migrated during the migration process, and can directly be used to enable block generation on the v4 node.

To directly set the values, use the `--use-status-value` flag, or set the values manually as described below.

For setting the values manually, first get the data:
[source,bash]
----
lisk-core generator:status --pretty
----

And then set the values manually by adding the relevant flags:
[source,bash]
----
lisk-core generator:enable lskqaxxmj78frvgpjgwvf4yqjjkcrr9yhn2sxxwm3 --height=123 --max-height-generated=101 --max-height-prevoted=101
----
====

=== Register legacy validator keys

.This step is crucial for migrating validators!

> IMPORTANT: As already mentioned, all non-initial validators will be banned during the bootstrap period.

Reviewer 3: Also, please highlight `banned` here 🙂

[tabs]
====
Mainnet::
+
--
[source,bash,subs=attributes+]
----
lisk-migrator --snapshot-height {snapshotHeight} --output ~/.lisk/lisk-core/config/mainnet --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
----
--
Testnet::
+
--
[source,bash,subs=attributes+]
----
lisk-migrator --snapshot-height {snapshotHeightTestnet} --output ~/.lisk/lisk-core/config/testnet --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
----
--
====

* `--snapshot-height`:
The height on which the blockchain snapshot will be performed.
The snapshot height will be announced separately.
* `--output`:
The absolute path to the directory, where the newly generated genesis block should be saved.

[TIP]
====
In case a custom xref:{url_sdk_config_system}[dataPath] is defined in the config, different to the default path `~/.lisk/lisk-core`, then it is possible to define the data path with the `--lisk-core-v3-data-path` flag like so:

[source,bash,subs=attributes+]
----

> lisk-migrator --snapshot-height {snapshotHeight} --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/.lisk/custom/path/ --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4

Reviewer 3: Maybe we should avoid the `.` in `--lisk-core-v3-data-path ~/lisk/custom/path/`.
```suggestion
lisk-migrator --snapshot-height {snapshotHeight} --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/lisk/custom/path/ --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
```

----
====

[TIP]
====
It is possible to use tools such as `screen` to run the Lisk migrator in the background.

With `screen` you can detach the current terminal window into the background:

.Example (Mainnet) footnote:snap_footnote[]
[source,bash,subs=attributes+]
----
screen -dmSL migration lisk-migrator --snapshot-height {snapshotHeight} --output ~/.lisk/lisk-core/config/mainnet --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
----

Shortly before the migration occurs, it is possible to reattach to the screen, in order to check if everything is working correctly.

First, check the name of the detached screen:

[source,bash]
----
screen -ls
----

This returns a list of all detached screens with `screen`:

----
There is a screen on:
	1842.migration	(05/07/2021 12:35:59 PM)	(Detached)
1 Socket in /run/screen/S-lisk.
----

Use `screen -r` and the name of the detached screen you want to connect to

[source,bash]
----
screen -r 1842.migration
----
====

=== Wait until the network reaches the snapshot height

> Observe if the `lisk-migrator` finishes successfully, (this can take several minutes).

Reviewer 3: It is probably better to give them an estimation. approx, 10 mins against the testnet and 3hrs against the mainnet.

> The flag `--auto-download-lisk-core-v4` will also download and install Lisk Core v4 for you.

Author: it wont only download the CLI tool, but whole Lisk Core, or not? So why update it?

> The flag `--auto-download-lisk-core-v4` will also download and install Lisk Core v4 for you.

Reviewer 3: No, it only downloads the Lisk Core CLI from npm registry. Check [here](https://github.com/LiskHQ/lisk-migrator/blob/00b953da7ee8d16dd80624cad6498a78c496677c/src/utils/node.ts#L31).

The flag `--auto-migrate-config` will automatically migrate the config from the old to the new node.

After the snapshot height is reached, validators have approximately {initRoundsTime} to enable block generation, to ensure they will not miss any blocks after the hard fork.

NOTE: If the node is not migrated, but started at a later point in time, it will simply sync to the current network height.

> For validators, this might result in missing blocks, for everyone else it will not have any impact.

Reviewer 3: ```suggestion
For validators, this might result in missing blocks. For everyone else, it will not have any impact.
```

After creating the config, start it with the following command:

./home/lisk/lisk-core/
[source,bash]
----
pm2 start pm2.conf.json
----

This will start Lisk Core in the background.

You can verify that the node is running correctly by executing the following command:

[source,bash]
----
lisk-core system:node-info --pretty
----

This will return certain general node information.

=== Stop Lisk Core v3.0.5

After the `lisk-migrator` script has finished and the announced snapshot height has passed, there is no reason to continue running Lisk Core v3.0.5, and therefore it is recommended to stop it.

//TODO: Update with correct command to stop Lisk Core
////
Navigate into the root folder of your Lisk Core v3.0.5 installation and run the following command to stop the old Lisk Core version:

[source,bash]
----
lisk-core
----
////

NOTE: Users that have set something up to start Lisk Core on boot, whether it's through `cron` or `systemd` or something else, should keep that in mind and adjust things accordingly so that Lisk Core 4.x gets started instead of Lisk Core 3.x.

Last but not least, remove the folder with Lisk Core v3.0.5, e.g. by executing the following:

[source,bash]
----

> rm -r lisk-core

Author: This is referring to the old `lisk-core` version, which i assume to be present under `lisk-core/`, but maybe it is not clear enough as the folder with new version might have same name. 🤔  Will remove the code snippet, to avoid confusion

----

=== Enable block generation after the migration (for validators)

After migration, *101 initial validators* will be active to generate blocks during the `initRounds`.

`initRounds` is the number of rounds for the bootstrap period of the new network.
The bootstrap period after migration to Core v4 is {initRounds} rounds.

The initial validators will be exactly the 101 validators that were in active positions in the Lisk Core v3 network, at the time of the migration.

For the initial validators, it will be important to enable block generation as soon as possible on the new node, to not miss any block rewards.

It is also important that as many as possible validators participate in the network migration.
If only a small number of validators migrate their nodes, this can prolong the bootstrap period for a long time.

> Additionally, for blocks to be finalized, enough of the active validators (min. 68) need to participate in the migration process.

Reviewer 3: Since we do not usually use abbreviations in our documentations.
```suggestion
Additionally, for blocks to be finalized, enough of the active validators (minimum 68) need to participate in the migration process.
```

.config.json
[source,json]
----
{
  "rpc": {
    "modes": ["ipc"],
    "allowedMethods": ["generator", "system", "random"]
  }
}
----

Restart the node with the `--overwrite-config` flag, to load the updated configuration.

[TIP]
====
`allowedMethods` refers to the `method` defined in the {url_jsonrpc}[JSON-RPC specification^].

Add the `namespace`, to allow all endpoints of `namespace`, or `namespace_endpointName` to allow a specific endpoint.
====

==== Create the validator keys

It is possible to generate all relevant validator keys to enable block generation from the account passphrase.

To do so, use the command xref:{url_sdk_cli_keyscreate}[keys:create].

[source,bash]
----
lisk-core keys:create --output config/keys.json
----

Next, you will be prompted for the validator *passphrase*, and it will also ask for a *password*, in order to symmetrically encrypt the passphrase for the config.

----
? Please enter passphrase:  [hidden]
? Please re-enter passphrase:  [hidden]
? Please enter password:  [hidden]
? Please re-enter password:  [hidden]
----

CAUTION: The password is sensitive information.
Store the password used here for the encryption somewhere safe.

> It will be required every time to enable block generation, in order to decrypt the generator keys in the node.

Reviewer: I would keep `The password is sensitive information` as is and only change 'in' to 'on'

> It will be required every time to enable block generation, in order to decrypt the generator keys in the node.

Reviewer 3: My bad 🤦 Updated the suggestion.

        "blsPrivateKey": "36506a53431665265ee03d7e19a5d44db3ff159d9aeee05727a8b24abc67651a"
      },
      "encrypted": {
        "ciphertext": "c3009d4a505ac32a652ffce6aa718073c7ca75b00578420ba20c2533a83f38e2b3e20cf1d6f0c9905efe28b5276142b93fdbdd33134d37bcd2db23654da92bb2becd00971c49ecc749100748c93344477ea52f6073c3fefec7234962d0eccdaa6862d9d0da46dbfe85cef98ad6cab0f2c1cb1b54326617132bb950d1c14a774a1e6403e8fa1bf3a2c7c0d6856266cf738f41ac01b2217d93070c4079e1b82044d3a692ea225290c2b6bcb902e0ffb8132f4c0f29325e6a3a",
        "mac": "2b3c65d0385a870ab499dfcddf411347506671015f412b35600153b132a455ea",
        "kdf": "argon2id",
        "kdfparams": {
          "parallelism": 4,
          "iterations": 1,
          "memorySize": 2024,
          "salt": "a5598628001346f608b3f57dd38b8611"
        },
        "cipher": "aes-256-gcm",
        "cipherparams": {
          "iv": "683600a199d154e51c0f97e6",
          "tag": "32807058f7f89921b4839fc39256cd24"
        },
        "version": "1"
      }
    }
  ]
}
----

[CAUTION]
====
The generated keys are very sensitive information.

> Especially the non encrypted values need to be treated *as equally sensitive as the passphrase for an account.*

Reviewer 3: ```suggestion
Especially the non-encrypted values need to be treated *as equally sensitive as the passphrase for an account.*
```

Now, it is possible to enable block generation on the new node for your validator by using the xref:{url_cli_generatorenable}[generator:enable] command of the Lisk Core CLI.

[source,bash]
----
lisk-core generator:enable lskqaxxmj78frvgpjgwvf4yqjjkcrr9yhn2sxxwm3 --use-status-value
----

Replace the address with your validator address.

.Don't use zeros as validator info data!
[WARNING]
====
If the validator already generated blocks with Lisk Core v3, they need to use their current validator info data.
The validator info data is migrated during the migration process, and can directly be used to enable block generation on the v4 node.

To directly set the values, use the `--use-status-value` flag, or set the values manually as described below.

For setting the values manually, first get the data:
[source,bash]
----
lisk-core generator:status --pretty
----

And then set the values manually by adding the relevant flags:
[source,bash]
----
lisk-core generator:enable lskqaxxmj78frvgpjgwvf4yqjjkcrr9yhn2sxxwm3 --height=123 --max-height-generated=101 --max-height-prevoted=101
----
====

=== Register legacy validator keys

.This step is crucial for migrating validators!

> IMPORTANT: The registration of the legacy validator keys has to be done *latest at the end of `initRounds`*, to be eligible to generate blocks in the new network.

Reviewer 3: It's important to highlight/repeat the same information about non-init validators being banned during the migration process.

Author: Is there anywhere, where you would you like this to be highlighted/repeated specifically?

Reviewer 3: In the `Register legacy validator keys` section as an important note perhaps!?

> Otherwise, the validator will be *banned* in the network, until the validator keys are registered in the legacy module.

Reviewer 3: ```suggestion
Otherwise, the validator will continue to be *banned* on the network, until the validator keys are registered in the legacy module.
```

= Testnet and Mainnet v4 migration guide
:toc: preamble
:experimental:
:idprefix:
:idseparator: -
//TODO: Update TBD values
:snapshotHeight: TBD
:snapshotHeightTestnet: TBD
:initRounds: TBD

> :initRoundsTime: TBD hours

Reviewer 3: `initRounds` will be `60480`. So, `initRoundsTime` will be approx 1 week.
Ref: `POS_INIT_ROUNDS` in https://github.com/LiskHQ/lips/blob/main/proposals/lip-0063.md#constants.

:url_wallet: https://lisk.com/wallet
:url_jsonrpc: https://www.jsonrpc.org/specification
//TODO: Update links to blog posts
:url_lisk_blog_migration: https://lisk.com/blog/development/announcing-lisk-testnet-v3-migration
:url_lisk_blog_migration_mainnet: https://lisk.com/blog/development/announcing-lisk-mainnet-v3-migration
// Project URLs
:url_run_validator: beta@ROOT::run-blockchain/become-validator.adoc
:url_sdk_cli_keyscreate: {docs_sdk}client-cli.adoc#keyscreate
:url_sdk_config_system: {docs_sdk}config.adoc#system
:url_cli_generatorenable: core-cli.adoc#generatorenable
:url_cli_endpointinvoke: core-cli.adoc#endpointinvoke

This guide explains how to migrate a Lisk Core v3.0.5 node to Lisk Core v4.x using the Lisk migrator v2.

The Lisk migrator CLI will generate a new genesis block for Lisk Core v4.
The new genesis block is created based on a snapshot of the old blockchain (running on Lisk Core v3.0.5) at a given height.

[IMPORTANT]

====
All active validators on the Lisk Mainnet and the Lisk Testnet need to follow this guide to correctly migrate their nodes to the new network, in order to not miss any blocks after the network hard fork.

> Optionally, anyone running a Lisk Core v3.0.5 who is not an active validator can participate in the migration process.

Reviewer 3: ```suggestion
Optionally, anyone running a Lisk Core v3.0.5 who is not an active validator can also participate in the migration process.
```

> In this case, the following steps 2.7 and 2.8 in this guide can be skipped, as they are only relevant for validators.

Reviewer 3: ```suggestion
In this case, the following steps 2.7 and 2.8 in this guide can be skipped, as they are only relevant for the current validators.
```

====

:sectnums:
:sectnumlevels: 5
== Preparation

.System requirements
[NOTE]

====
The following system requirements are recommended for running the migration script:

Memory::

> * minimum of 4 GB RAM

Reviewer 3: We have a minimum requirement of 4 GB RAM for only Lisk Core. So, to run the migrator we will ideally need more but I do not have numbers around that. Maybe @Nazgolze can suggest a safe number? (6 or 8 GB perhaps).

Reviewer: That's a good point.

It's the genesis block that takes up a lot of RAM. I can't really say what a good number is without testing this

Reviewer 3: Then, we should perhaps leave a TODO so that we can come back and update with the exact recommendation. @Tschakki For now, I'd recommend you to update it to 8GB instead.

OS::
* Ubuntu 20.04 LTS
* Ubuntu 22.04 LTS

Node.js::
* Version 18 (LTS)
====

=== Ensure you are running version 3.0.5 of Lisk Core
Ensure you are running version 3.0.5 of Lisk Core to be able to seamlessly migrate to Lisk Core 4.0.0.

Navigate into the root folder of your Lisk Core installation, and open `package.json` to check the `version` of your Lisk Core node.

> If your version is lower than 3.0.5, ensure to update Lisk Core before proceeding with the migration.

Reviewer 3: Users should preferrably check the node info response for the version of their running node (`lisk-core node:info --pretty`). If their HTTP API is enabled, they can check `data.version` from `https://mainnet.lisk.com/api/node/info` (reference).
Or, `apiClient.node.getNodeInfo();`

----
{
  "name": "lisk-core",
  "script": "lisk-core start",
  "env": {
    "LISK_NETWORK": "mainnet"
  }
}
----
--
Testnet::
+
--
.~/lisk-core/pm2.conf.json
[source,json]
----
{
  "name": "lisk-core",
  "script": "lisk-core start",
  "env": {
    "LISK_NETWORK": "testnet"
  }
}
----
--
====

TIP: All available options for `scripts` and `env` can be found in `lisk-core start --help`.

After creating the config, start it with the following command:

./home/lisk/lisk-core/
[source,bash]
----
pm2 start pm2.conf.json
----

This will start Lisk Core in the background.

You can verify that the node is running correctly by executing the following command:

[source,bash]
----
lisk-core system:node-info --pretty
----

This will return certain general node information.

=== Stop Lisk Core v3.0.5

After the `lisk-migrator` script has finished and the announced snapshot height has passed, there is no reason to continue running Lisk Core v3.0.5 and therefore it is recommended to stop it.

First ensure, that Lisk Core will not start again by removing the existing cronjob:

[source,bash]
----
crontab -e
----

Choose your favorite editor to open the `crontab` file and remove the following line:

>  @reboot /bin/bash /home/lisk/lisk-test/lisk.sh start > /home/lisk/lisk-test/cron.log 2>&1

Reviewer: This is old, from the lisk 2.x and older days, and should be removed.

Users that have set something up to start lisk-core on boot, whether it's through cron or systemd or something else, should keep that in mind and adjust things accordingly so that lisk-core 4.x gets started instead of lisk-core 3.x

Save and close the `crontab` file again.

Navigate into the root folder of your Lisk Core v3.0.5 installation and run the following command to stop the old Lisk Core version:

[source,bash]
----

> bash lisk.sh stop

Reviewer: This is old too. Anything referencing `lisk.sh` should be changed or removed

Author: @Nazgolze what is the command for this in the new version?

Compare the current height of your node to the network height in {url_wallet}[Lisk Desktop^], which is shown on the kbd:[Network] or kbd:[Blocks] pages.

TIP: To view the current height of the *Lisk Testnet*, use the network switcher of Lisk Desktop, which can be enabled in the settings.

> Alternatively users can also verify the current height by comparing `data.height` in the response from the https://service.lisk.com/api/v2/network/status endpoint.

Reviewer 2: ```suggestion
Alternatively, users can also verify the current height by comparing `data.height` in the response from the https://service.lisk.com/api/v2/network/status endpoint.
```

> To directly check the current height via command line, run:

Reviewer 2: ```suggestion
To directly check the current height via the command line, run:
```
